pallet/staking/src/lib.rs

```diff
@@ -735,215 +827,276 @@ pub mod pallet {
+			} else {
+				Self::deposit_event(Event::Unpaid { staker: collator, amount: c_payout });
+			}
+
+			Ok(<T as Config>::WeightInfo::payout())
 		}
 
 		/// Prepare the session state.
-		pub fn prepare_new_session() {
-			<RewardPoints<T>>::kill();
-			#[allow(deprecated)]
-			<Exposures<T>>::remove_all(None);
-			<NextExposures<T>>::iter().drain().for_each(|(k, v)| {
-				<Exposures<T>>::insert(k, v);
-			});
+		pub fn prepare_new_session(index: u32) -> Option<Vec<T::AccountId>> {
+			<Pallet<T>>::shift_exposure_cache_states();
+
+			log::info!(
+				"[pallet::staking] assembling new collators for new session {index} at #{:?}",
+				<frame_system::Pallet<T>>::block_number(),
+			);
+
+			if let Ok(collators) = Self::elect() {
+				// TODO?: if we really need this event
+				Self::deposit_event(Event::Elected { collators: collators.clone() });
+
+				Some(collators)
+			} else {
+				// Impossible case.
+				//
+				// But if there is an issue, retain the old collators; do not alter the session
+				// collators if any error occurs to prevent the chain from stalling.
+				None
+			}
+		}
+
+		/// Shift the exposure cache states.
+		///
+		/// Previous Current  Next
+		/// Next     Previous Current
+		/// Current  Next     Previous
+		///
+		/// ```nocompile
+		/// loop { mutate(2, 0, 1) }
+		/// ```
+		pub fn shift_exposure_cache_states() {
+			let (s0, s1, s2) = <ExposureCacheStates<T>>::get();
+
+			<ExposureCacheStates<T>>::put((s2, s0, s1));
 		}
 
 		/// Elect the new collators.
 		///
 		/// This should only be called by the [`pallet_session::SessionManager::new_session`].
-		pub fn elect() -> Vec<T::AccountId> {
+		pub fn elect() -> Result<Vec<T::AccountId>, DispatchError> {
+			let nominators = <Nominators<T>>::iter().collect::<Vec<_>>();
+			let ring_pool = <RingPool<T>>::get();
+			let kton_pool = <KtonPool<T>>::get();
 			let mut collators = <Collators<T>>::iter()
 				.map(|(c, cm)| {
 					let scaler = Perbill::one() - cm;
 					let mut collator_v = 0;
-					let nominators = <Nominators<T>>::iter()
+					let nominators = nominators
+						.iter()
 						.filter_map(|(n, c_)| {
-							if c_ == c {
-								let nominator_v = scaler * Self::power_of(&n);
+							if c_ == &c {
+								let nominator_v = scaler * Self::power_of(n, ring_pool, kton_pool);
 
 								collator_v += nominator_v;
 
-								Some(IndividualExposure { who: n, vote: nominator_v })
+								Some(IndividualExposure { who: n.to_owned(), vote: nominator_v })
 							} else {
 								None
 							}
 						})
 						.collect();
 
-					((c, Exposure { vote: collator_v, nominators }), collator_v)
+					((c, Exposure { commission: cm, vote: collator_v, nominators }), collator_v)
 				})
 				.collect::<Vec<_>>();
 
 			collators.sort_by(|(_, a), (_, b)| b.cmp(a));
 
+			let cache_states = <ExposureCacheStates<T>>::get();
+
+			#[allow(deprecated)]
+			call_on_exposure!(cache_states, Next::remove_all(None))?;
```

Comment:
Suggest moving this out to the `prepare_new_session` after `shift_exposure_cache_states`